src/fastgltf_tools.hpp

```diff
@@ -0,0 +1,467 @@
+};
+
+template<>
+struct ElementTraits<std::uint16_t> {
+	using element_type = std::uint16_t;
+	using component_type = std::uint16_t;
+	static constexpr auto type = AccessorType::Scalar;
+	static constexpr auto enum_component_type = ComponentType::UnsignedShort;
+};
+
+template<>
+struct ElementTraits<std::int32_t> {
+	using element_type = std::int32_t;
+	using component_type = std::int32_t;
+	static constexpr auto type = AccessorType::Scalar;
+	static constexpr auto enum_component_type = ComponentType::Int;
+};
+
+template<>
+struct ElementTraits<std::uint32_t> {
+	using element_type = std::uint32_t;
+	using component_type = std::uint32_t;
+	static constexpr auto type = AccessorType::Scalar;
+	static constexpr auto enum_component_type = ComponentType::UnsignedInt;
+};
+
+template<>
+struct ElementTraits<float> {
+	using element_type = float;
+	using component_type = float;
+	static constexpr auto type = AccessorType::Scalar;
+	static constexpr auto enum_component_type = ComponentType::Float;
+};
+
+template<>
+struct ElementTraits<double> {
+	using element_type = double;
+	using component_type = double;
+	static constexpr auto type = AccessorType::Scalar;
+	static constexpr auto enum_component_type = ComponentType::Double;
+};
+
+#if FASTGLTF_HAS_CONCEPTS
+template <typename ElementType>
+concept Element = std::is_arithmetic_v<typename ElementTraits<ElementType>::component_type>
+		&& ElementTraits<ElementType>::type != AccessorType::Invalid
+		&& ElementTraits<ElementType>::enum_component_type != ComponentType::Invalid
+		&& std::is_default_constructible_v<ElementType>
+		&& std::is_constructible_v<ElementType>
+		&& std::is_move_assignable_v<ElementType>;
+#endif
+
+namespace internal {
+
+template <typename SourceType, typename DestType, std::size_t Index>
+constexpr DestType convertComponent(const std::byte* bytes) {
+	return static_cast<DestType>(reinterpret_cast<const SourceType*>(bytes)[Index]);
+}
+
+template <typename ElementType, typename SourceType, std::size_t... I>
+#if FASTGLTF_HAS_CONCEPTS
+requires Element<ElementType>
+#endif
+constexpr ElementType convertAccessorElement(const std::byte* bytes, std::index_sequence<I...>) {
+	using DestType = typename ElementTraits<ElementType>::component_type;
+	static_assert(std::is_arithmetic_v<DestType>, "Accessor traits must provide a valid component ttype");
+
+	if constexpr (std::is_aggregate_v<ElementType>) {
+		return {convertComponent<SourceType, DestType, I>(bytes)...};
+	} else {
+		return ElementType{convertComponent<SourceType, DestType, I>(bytes)...};
+	}
+}
+
+template <typename ElementType,
+		typename Seq = std::make_index_sequence<getNumComponents(ElementTraits<ElementType>::type)>>
+#if FASTGLTF_HAS_CONCEPTS
+requires Element<ElementType>
+#endif
+ElementType getAccessorElementAt(ComponentType componentType, const std::byte* bytes) {
+	switch (componentType) {
+		// This is undefined behavior if component type is invalid
+		default:
```

Comment:
I seriously think we shouldn't have a `default` branch here. If I were to add a new component type, the compiler would warn me about a missing branch here. With the `default` branch, we gain nothing but catching UB (casting to a enum class with a non-existent value) that isn't even caused by us. See this:
> The static_cast operator can explicitly convert an integral value to an enumeration type. If the value of the integral type does not fall within the range of enumeration values, the resulting enumeration value is undefined.

Reply:
Fixed in [df17edc](https://github.com/spnda/fastgltf/pull/20/commits/df17edc742f1893ece80c0cc2467a9b68b7ec880)